polar-core/src/vm.rs

```diff
@@ -1926,6 +1931,74 @@ impl PolarVirtualMachine {
         Ok(QueryEvent::None)
     }
 
+    /// Check for partially bound or unbound variable arguments to an external call.
+    /// If any arguments are partially bound or unbound, return an error.
+    /// The only exception is the method `OsoRoles.role_allows()`, which expects a partially bound resource argument
+    fn check_partial_args(
+        &mut self,
+        object: &Term,
+        field: &Term,
+        value: &Term,
+    ) -> PolarResult<bool> {
+        // If the lookup is a `Value::Call`, then we need to check for partial args
+        let mut has_partial_arg = false;
+        if let Value::Call(Call { name, args, kwargs }) = field.value() {
+            // get all arg values (args + kwargs)
+            let mut arg_values = args.clone();
+            if let Some(kwarg_map) = kwargs.clone() {
+                let mut v: Vec<Term> = kwarg_map.values().cloned().collect();
+                arg_values.append(&mut v)
+            }
+            for (i, arg) in arg_values.iter().enumerate() {
+                // Partial/unbound variables cannot be passed to external methods from Polar
+                if let Value::Variable(v) = arg.value() {
+                    match self.binding_manager.variable_state(v) {
+                        // bound variables are fine, continue
+                        VariableState::Bound(_) => continue,
+                        // partial variables are only fine if being passed into "role_allows"
+                        VariableState::Partial => {
+                            if let Value::ExternalInstance(external) =
+                                self.deep_deref(&object).value()
+                            {
+                                if let Some(repr) = external.repr.clone() {
+                                    // only allow the third argument (resource) to be partially bound
+                                    if repr.contains("sqlalchemy_oso.roles2.OsoRoles")
+                                        && name.0 == "role_allows"
+                                        && i == 2
+                                    {
+                                        // add constraint for role_allows call
+                                        let dot = op!(
+                                            Dot,
+                                            object.clone(),
+                                            field.clone_with_value(Value::Call(Call {
+                                                name: name.clone(),
+                                                args: vec![arg.clone()],
+                                                kwargs: None
+                                            }))
+                                        );
+                                        let term =
+                                            &op!(Unify, value.clone(), dot.into_term()).into_term();
+                                        self.add_constraint(term)?;
```

Comment:
It's unexpected from the caller that this code adds a constraint. I'd either rename the function & update the doc string to make it more clear that a constraint will be added, or return an `Option<Term>` which contains the constraint to add, adding it in the calling code.

Reply:
@leina05 this one is the biggest deal of my comments tbh